osquery/core/windows/kobjhandle.cpp

```diff
@@ -0,0 +1,123 @@
+/**
+ *  Copyright (c) 2014-present, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under both the Apache 2.0 license (found in the
+ *  LICENSE file in the root directory of this source tree) and the GPLv2 (found
+ *  in the COPYING file in the root directory of this source tree).
+ *  You may select, at your option, one of the above-listed licenses.
+ */
+
+#define _WIN32_DCOM
+
+#include "Kobjhandle.h"
```

Comment:
Not sure we need the `k` in the header and impl file/class names? Why not just `objhandle` or `objecthandle`?

Reply:
Looks like once you have this as a relative include path the build should at least get beyond it's current error.

Reply:
@muffins the reason for the k (for kernel) was an attempt to limit the scope of the class to _just_ kernel objects (as compared to non-kernel objects such as services (SC_HANDLE) etc.  I have _zero_ issue with changing it, just let me know.

the full path on includes is done.  it's building locally without the full path, so I guess I don't understand why need the full path, but no problem obviously changing.

Reply:
The biggest observation I'd make about that is not all HANDLE "invalid values" are created equally. Some use NULL, some use INVALID_HANDLE. In practice, I believe they're the same but that's not a good thing to rely on :).

Reply:
@zacbrown for sure.  INVALID_HANDLE_VALUE is (unsigned long)-1.  NULL is of course 0.

INVALID_HANDLE_VALUE is returned for all fs objects (files, sockets, pipes).  NULL for evertyhig else.

Another key difference is the use of CloseHandle.  For example, even though a HDESKTOP is, in fact, a kernel HANDLE, CloseDesktop() provides some custom logic specific to desktops.  If CloseHandle() is used on a handle to a desktop, it raises an exception.  There are a series of other cases, including some in reverse.  For example, a HKEY is a kernel handle except when it isn't - which includes remote registry psuedo-handles.  If the HKEY references a HANDLE opened to the remote registry service, even if the remeote registry service is on the local windows instance, then of course the HKEY provides a reference to an object on the remote endpoint.

Net it out - absolutely get it.  My hope was that this class could provide the basis of future abstractions.

Reply:
@scottlundgren that's fine, I'm definitely down to leave the class, and re: the `k` part, I'd say what about changing the file name to `kernel_object_handles.cpp`? For something like a helper/utility class I think it'd be better to have naming be a bit more explicit, that being said I'd say let's stick with the class, you have compelling points :)

Reply:
renamed just handle.[h|cpp] and renamed the class Handle.  Let's see how that looks to someone else's eyes, figured after all my yakking about HKEYs that it would actually be helfpul to abstract away HANDLEs taht aren't really kernel handles so anyone else doesn't have to think about it.